include/albatross/src/core/linear_combination.hpp

```diff
@@ -74,6 +74,63 @@ inline auto difference(const X &x, const Y &y) {
   return difference(vx, vy);
 }
 
+template <typename Derived, typename X>
+inline auto operator*(const Eigen::SparseMatrixBase<Derived> &matrix,
+                      const std::vector<X> &features) {
+  std::vector<LinearCombination<X>> output(cast::to_size(matrix.rows()));
+
+  std::vector<std::vector<double>> output_coefs(cast::to_size(matrix.rows()));
+
+  for (Eigen::Index k = 0; k < matrix.outerSize(); ++k) {
+    for (typename Derived::InnerIterator it(matrix.derived(), k); it; ++it) {
+      output[cast::to_size(it.row())].values.push_back(
+          features[cast::to_size(it.col())]);
+      output_coefs[cast::to_size(it.row())].push_back(it.value());
+    }
+  }
+
+  for (std::size_t i = 0; i < output_coefs.size(); ++i) {
+    output[i].coefficients = Eigen::Map<Eigen::VectorXd>(
+        &output_coefs[i][0], cast::to_index(output_coefs[i].size()));
+  }
+
+  return output;
+}
+
+template <typename Derived, typename X>
+inline auto operator*(const Eigen::SparseMatrixBase<Derived> &matrix,
+                      const std::vector<LinearCombination<X>> &features) {
+  std::vector<LinearCombination<X>> output(cast::to_size(matrix.rows()));
+
+  std::vector<std::vector<double>> output_coefs(cast::to_size(matrix.rows()));
+
+  for (Eigen::Index k = 0; k < matrix.outerSize(); ++k) {
+    for (typename Derived::InnerIterator it(matrix.derived(), k); it; ++it) {
+      const auto &combo = features[cast::to_size(it.col())];
+      for (Eigen::Index i = 0; i < combo.coefficients.size(); ++i) {
+        output[cast::to_size(it.row())].values.push_back(
```

Comment:
This statement doesn't require iteratively pushing, right?  can't we just copy `combo.values` outside this loop?

Reply:
This passes all tests for me:

```
@@ -102,26 +102,16 @@ inline auto operator*(const Eigen::SparseMatrixBase<Derived> &matrix,
                       const std::vector<LinearCombination<X>> &features) {
   std::vector<LinearCombination<X>> output(cast::to_size(matrix.rows()));
 
-  std::vector<std::vector<double>> output_coefs(cast::to_size(matrix.rows()));
-
   for (Eigen::Index k = 0; k < matrix.outerSize(); ++k) {
     for (typename Derived::InnerIterator it(matrix.derived(), k); it; ++it) {
       const auto row = cast::to_size(it.row());
       const auto col = cast::to_size(it.col());
       const auto &combo = features[col];
-      for (Eigen::Index i = 0; i < combo.coefficients.size(); ++i) {
-        const auto si = cast::to_size(i);
-        output[row].values.push_back(combo.values[si]);
-        output_coefs[row].push_back(it.value() * combo.coefficients[i]);
-      }
+      output[row].values = combo.values;
+      output[row].coefficients = it.value() * combo.coefficients;
     }
   }
 
-  for (std::size_t i = 0; i < output_coefs.size(); ++i) {
-    output[i].coefficients = Eigen::Map<Eigen::VectorXd>(
-        &output_coefs[i][0], static_cast<Eigen::Index>(output_coefs[i].size()));
-  }
-
   return output;
 }
```